File: Lernwerk/Services/LLM/ClaudeClient.swift
import Foundation

/// Calls the Claude Messages API over plain HTTP; there is no official Swift SDK.
struct ClaudeClient: LLMClient {
    static let endpoint = URL(string: "https://api.anthropic.com/v1/messages")!
    static let apiVersion = "2023-06-01"
    static let fallbackBeta = "server-side-fallback-2026-07-01"

    var apiKey: String
    var model: String
    var session: URLSession = .shared

    func complete(_ request: LLMRequest) async throws -> LLMResponse {
        var urlRequest = URLRequest(url: Self.endpoint)
        urlRequest.httpMethod = "POST"
        urlRequest.timeoutInterval = 600
        urlRequest.setValue("application/json", forHTTPHeaderField: "content-type")
        urlRequest.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        urlRequest.setValue(Self.apiVersion, forHTTPHeaderField: "anthropic-version")
        if Self.supportsFallbacks(model) {
            urlRequest.setValue(Self.fallbackBeta, forHTTPHeaderField: "anthropic-beta")
        }
        urlRequest.httpBody = try JSONSerialization.data(withJSONObject: Self.body(for: request, model: model))

        let (data, response) = try await session.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else {
            throw LLMError.invalidResponse
        }
        return try Self.parse(data: data, status: http.statusCode, expectsJSON: request.jsonSchema != nil)
    }

    static func body(for request: LLMRequest, model: String) -> [String: Any] {
        var body: [String: Any] = [
            "model": model,
            "max_tokens": request.maxTokens,
            "system": request.system,
            "messages": request.messages.map(encodeMessage),
        ]

        var outputConfig: [String: Any] = [:]
        if let effort = request.effort, supportsEffort(model) {
            outputConfig["effort"] = effort.rawValue
        }
        if let schema = request.jsonSchema {
            let format: [String: Any] = ["type": "json_schema", "schema": schema]
            outputConfig["format"] = format
        }
        if !outputConfig.isEmpty {
            body["output_config"] = outputConfig
        }
        if supportsFallbacks(model) {
            body["fallbacks"] = "default"
        }
        return body
    }

    static func encodeMessage(_ message: LLMMessage) -> [String: Any] {
        [
            "role": message.role.rawValue,
            "content": message.content.map(encodeContent),
        ]
    }

    static func encodeContent(_ content: LLMContent) -> [String: Any] {
        switch content {
        case let .text(text):
            return ["type": "text", "text": text]
        case let .image(jpeg):
            let source: [String: Any] = [
                "type": "base64",
                "media_type": "image/jpeg",
                "data": jpeg.base64EncodedString(),
            ]
            return ["type": "image", "source": source]
        case let .pdf(data):
            let source: [String: Any] = [
                "type": "base64",
                "media_type": "application/pdf",
                "data": data.base64EncodedString(),
            ]
            return ["type": "document", "source": source]
        }
    }

    static func supportsFallbacks(_ model: String) -> Bool {
        model == "claude-opus-5" || model == "claude-fable-5-1"
    }

    static func supportsEffort(_ model: String) -> Bool {
        model.hasPrefix("claude-opus") || model.hasPrefix("claude-sonnet-5") || model.hasPrefix("claude-fable")
    }

    static func parse(data: Data, status: Int, expectsJSON: Bool) throws -> LLMResponse {
        let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]

        guard (200..<300).contains(status) else {
            let apiError = object?["error"] as? [String: Any]
            let message = apiError?["message"] as? String ?? String(decoding: data, as: UTF8.self)
            throw LLMError.http(status: status, message: message)
        }
        guard let object, let blocks = object["content"] as? [[String: Any]] else {
            throw LLMError.invalidResponse
        }

        let stopReason = object["stop_reason"] as? String
        if stopReason == "refusal" {
            throw LLMError.refusal
        }

        let text = blocks
            .filter { $0["type"] as? String == "text" }
            .compactMap { $0["text"] as? String }
            .joined()

        if stopReason == "max_tokens", expectsJSON || text.isEmpty {
            throw LLMError.truncated
        }
        guard !text.isEmpty else {
            throw LLMError.invalidResponse
        }
        return LLMResponse(text: text, stopReason: stopReason, model: object["model"] as? String)
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/llm/ClaudeClient.kt
package de.maxifrz.lernwerk.llm

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import java.util.Base64

/** Calls the Claude Messages API over plain HTTP. */
class ClaudeClient(
    private val apiKey: String,
    private val model: String,
    private val transport: HttpTransport,
) : LlmClient {
    override val capabilities = LlmCapabilities(acceptsImages = true, documentHandling = DocumentHandling.NATIVE_PDF)

    override suspend fun complete(request: LlmRequest): LlmResponse {
        val headers = buildMap {
            put("content-type", "application/json")
            put("x-api-key", apiKey)
            put("anthropic-version", API_VERSION)
            if (supportsFallbacks(model)) put("anthropic-beta", FALLBACK_BETA)
        }
        val result = transport.post(ENDPOINT, headers, body(request, model).toString(), request.purpose.timeoutSeconds)
        return parse(result.body, result.status, expectsJson = request.jsonSchema != null)
    }

    companion object {
        const val ENDPOINT = "https://api.anthropic.com/v1/messages"
        const val API_VERSION = "2023-06-01"
        const val FALLBACK_BETA = "server-side-fallback-2026-07-01"

        fun body(request: LlmRequest, model: String): JsonObject = buildJsonObject {
            put("model", model)
            put("max_tokens", request.maxTokens)
            put("system", request.system)
            put("messages", JsonArray(request.messages.map(::encodeMessage)))

            val outputConfig = buildJsonObject {
                if (request.effort != null && supportsEffort(model)) put("effort", request.effort.wire)
                request.jsonSchema?.let { schema ->
                    put("format", buildJsonObject {
                        put("type", "json_schema")
                        put("schema", schema)
                    })
                }
            }
            if (outputConfig.isNotEmpty()) put("output_config", outputConfig)
            if (supportsFallbacks(model)) put("fallbacks", "default")
        }

        fun encodeMessage(message: LlmMessage): JsonObject = buildJsonObject {
            put("role", message.role.wire)
            put("content", JsonArray(message.content.map(::encodeContent)))
        }

        fun encodeContent(content: LlmContent): JsonObject = when (content) {
            is LlmContent.Text -> buildJsonObject {
                put("type", "text")
                put("text", content.text)
            }
            is LlmContent.Image -> buildJsonObject {
                put("type", "image")
                put("source", base64Source("image/jpeg", content.jpeg))
            }
            is LlmContent.Pdf -> buildJsonObject {
                put("type", "document")
                put("source", base64Source("application/pdf", content.data))
            }
        }

        private fun base64Source(mediaType: String, data: ByteArray) = buildJsonObject {
            put("type", "base64")
            put("media_type", mediaType)
            put("data", Base64.getEncoder().encodeToString(data))
        }

        fun supportsFallbacks(model: String) = model == "claude-opus-5" || model == "claude-fable-5-1"

        fun supportsEffort(model: String) =
            model.startsWith("claude-opus") || model.startsWith("claude-sonnet-5") || model.startsWith("claude-fable")

        fun parse(body: String, status: Int, expectsJson: Boolean): LlmResponse {
            val obj = parseObject(body)
            if (status !in 200..299) {
                val message = obj?.obj("error")?.string("message") ?: body
                throw LlmError.Http(status, message)
            }
            val blocks = (obj?.get("content") as? JsonArray) ?: throw LlmError.InvalidResponse

            val stopReason = obj.string("stop_reason")
            if (stopReason == "refusal") throw LlmError.Refusal

            val text = blocks.mapNotNull { it as? JsonObject }
                .filter { it.string("type") == "text" }
                .mapNotNull { it.string("text") }
                .joinToString("")

            if (stopReason == "max_tokens" && (expectsJson || text.isEmpty())) throw LlmError.Truncated
            if (text.isEmpty()) throw LlmError.InvalidResponse
            return LlmResponse(text, stopReason, obj.string("model"))
        }
    }
}

internal fun parseObject(body: String): JsonObject? =
    runCatching { Json.parseToJsonElement(body).jsonObject }.getOrNull()

internal fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

internal fun JsonObject.obj(key: String): JsonObject? = this[key] as? JsonObject

internal fun JsonElement?.intOrNull(): Int? = (this as? JsonPrimitive)?.contentOrNull?.toDoubleOrNull()?.toInt()
